### PR TITLE
docs(ChatMessage): prototype of message reactions with Popup

### PR DESCRIPTION
**Pull request by mnajdova**
This PR adds prototype of how the reactions can be used with Popup, to display the users who reacted on the message. On hovering of the reaction, users can see the list which appears in Popup.

image

All prototypes related to the chat messages are now moved under one prototype:
image

### CODE REVIEW

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 1]
```diff
@@ -405,6 +405,13 @@ class Sidebar extends React.Component<any, any> {
         to: '/important-and-mention-messages',
         styles: menuItemStyles,
       },
+      {
+        key: 'message-reactions-with-popup',
```
Comment (bmdalex) [inline docs/src/components/Sidebar/Sidebar.tsx]: sidebar becomes pretty full of prototypes, can we try grouping chat related prototypes by context the same we are doing for:

- dropdowns: http://localhost:8080/prototype-dropdowns
- popups: http://localhost:8080/prototype-popups

**Review by bmdalex — APPROVED** [review]
code looks good 👍  but I have to say the design is really ugly... 😿 
one comment

File: docs/src/components/Sidebar/Sidebar.tsx (hunk 1 above)
Comment (mnajdova) [reply docs/src/components/Sidebar/Sidebar.tsx]: Will do, for design yep, we decided because of accessibility reasons to replace the List with Menu, so it is ugly at this moment. Will address it, thanks!

File: packages/react/src/components/Popup/Popup.tsx [hunk 2]
```diff
@@ -227,7 +231,7 @@ export default class Popup extends AutoControlledComponent<ReactProps<PopupProps
         {this.state.open &&
           Popup.isBrowserContext &&
           popupContent &&
-          ReactDOM.createPortal(popupContent, document.body)}
+          (inline ? popupContent : ReactDOM.createPortal(popupContent, document.body))}
```
Comment (layershifter) [inline packages/react/src/components/Popup/Popup.tsx]: Can we move this change to a separate PR? And cover it with UT

Comment (mnajdova) [reply packages/react/src/components/Popup/Popup.tsx]: Sure thing, I've even added new FocusZone component.. All of these component related changes will be moved to separate PRs, once the accessibility team confirms that this work as expected.

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    this.props.onFocusCapture && this.props.onFocusCapture(ev)
+    if (ev.target !== ev.currentTarget) {
+      // every time focus changes within the trap zone, remember the focused element so that
+      // it can be restored if focus leaves the pane and returns via keystroke (i.e. via a call to this.focus(true))
+      this._previouslyFocusedElementInTrapZone = ev.target as HTMLElement
+    }
+  }
+
+  private _onKeyboardHandler = (ev: React.KeyboardEvent<HTMLDivElement>): void => {
+    this.props.onKeyDown && this.props.onKeyDown(ev)
```
Comment (sophieH29) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: We might not need this,  as `AutoFocusZone` should just focus inner element on mount, it should be still possible to TAB out from it

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      ev.preventDefault()
+      ev.stopPropagation()
+    } else if (!ev.shiftKey && _lastTabbableChild === ev.target) {
+      focusAsync(_firstTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    }
+  }
+
+  private _forceFocusInTrap = (ev: Event, triggeredElement: HTMLElement) => {
```
Comment (sophieH29) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: We might not need this,  as `AutoFocusZone` should just focus inner element on mount, it should be still possible to TAB out from it

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    }
+  }
+
+  private _handleOutsideFocus = (ev: FocusEvent): void => {
+    const focusedElement = document.activeElement as HTMLElement
+    focusedElement && this._forceFocusInTrap(ev, focusedElement)
+  }
+
+  private _handleOutsideClick = (ev: MouseEvent): void => {
+    const clickedElement = ev.target as HTMLElement
```
Comment (sophieH29) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: We might not need this,  as `AutoFocusZone` should just focus inner element on mount, it should be still possible to TAB out from it

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        inline
+        content={{
+          content: (
+            <Ref
+              innerRef={(popupNode: HTMLElement) => {
+                this.popup = popupNode
+              }}
+            >
+              <AutoFocusZone firstFocusableSelector={'ui-menu__item:first-child'}>
```
Comment (sophieH29) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: by default it will try to focus first focusable element in zone, so it should work without specifying `firstFocusableSelector` of first menu item

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                <Menu
+                  items={['Jane Doe', 'John Doe']}
+                  vertical
+                  data-is-focusable={true}
+                  variables={{ borderColor: 'transparent' }}
+                  onKeyDown={this.handleKeyDownOnMenu}
+                />
+              </AutoFocusZone>
```
Comment (sophieH29) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: I think better to add `AutoFocusZone` through accessibility behavior (as we done with `FocusTrapZone`). Because if user use `popupFocusTrapBehvaior` and `AutoFocusZone`, there will be 2 zones rendered with overlapping logic. We should limit user's actions to using either `FocusTrapZone` or `AutoFocusZone`

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+  focusAsync,
+} from './focusUtilities'
+
+import { AutoFocusZoneProps } from './AutoFocusZone.types'
+import getUnhandledProps from '../../getUnhandledProps'
+import getElementType from '../../getElementType'
+import * as customPropTypes from '../../customPropTypes'
+
+/** AutoFocusZone is used to focus the first element inside it's children */
+export class AutoFocusZone extends React.Component<AutoFocusZoneProps, {}> {
```
Comment (sophieH29) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: `AutoFocusZone` has common logic with `FocusTrapZone` - focusing inner element on componentDidMount(). `FocusTrapZone` can extend `AutoFocusZone`'s props and reuse that logic. So basically we'll get
```
<FocusTrapZone>
   <AutoFocusZone>
   </AutoFocusZone>
</FocusTrapZone>
```

Note: we need only one `<div>` rendered

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+      // focus on the last item that had focus in the zone before we left the zone
+      focusAsync(this._previouslyFocusedElementInTrapZone)
+      return
+    }
+
+    const focusSelector =
+      firstFocusableSelector &&
+      (typeof firstFocusableSelector === 'string'
+        ? firstFocusableSelector
+        : firstFocusableSelector())
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```diff
-   const focusSelector =
-      firstFocusableSelector &&
-      (typeof firstFocusableSelector === 'string'
-        ? firstFocusableSelector
-        : firstFocusableSelector())
+ const focusSelector = callable(firstFocusableSelector)
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      <ElementType
+        {...unhandledProps}
+        className={className}
+        ref={this.createRef}
+        aria-labelledby={ariaLabelledBy}
+        onFocusCapture={this._onFocusCapture}
+      >
+        {this.props.children}
+      </ElementType>
+    )
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```diff
-      <ElementType
-        {...unhandledProps}
-        className={className}
-        ref={this.createRef}
-        aria-labelledby={ariaLabelledBy}
-        onFocusCapture={this._onFocusCapture}
-      >
-        {this.props.children}
-      </ElementType>
-    )
+<Ref innerRef={this.elementRef}>
+      <ElementType
+        {...unhandledProps}
+        className={className}
+        aria-labelledby={ariaLabelledBy}
+        onFocusCapture={this._onFocusCapture}
+      >
+        {this.props.children}
+      </ElementType>
+    )
+</Ref>
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+import { getNextElement, focusAsync } from './focusUtilities'
+
+import { AutoFocusZoneProps } from './AutoFocusZone.types'
+import getUnhandledProps from '../../getUnhandledProps'
+import getElementType from '../../getElementType'
+import * as customPropTypes from '../../customPropTypes'
+
+/** AutoFocusZone is used to focus the first element inside it's children */
+export class AutoFocusZone extends React.Component<AutoFocusZoneProps, {}> {
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```suggestion
export class AutoFocusZone extends React.Component<AutoFocusZoneProps> {
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  private _root: { current: HTMLElement | null } = { current: null }
+  private _previouslyFocusedElementOutsideTrapZone: HTMLElement
+  private _previouslyFocusedElementInTrapZone?: HTMLElement
+
+  private createRef = elem => (this._root.current = ReactDOM.findDOMNode(elem) as HTMLElement)
+
+  static propTypes = {
+    as: customPropTypes.as,
+    className: PropTypes.string,
+    ariaLabelledBy: PropTypes.string,
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: `ariaLabelledBy` and `className` should be unhandled props

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    firstFocusableSelector: PropTypes.string,
+    disableFirstFocus: PropTypes.bool,
+    focusPreviouslyFocusedInnerElement: PropTypes.bool,
+  }
+
+  static defaultProps: AutoFocusZoneProps = {
+    as: 'div',
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: Doesn't make any sense, `as` defaults to `div`

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 14]
```diff
@@ -20,51 +12,34 @@ import * as customPropTypes from '../../customPropTypes'
 
 /** AutoFocusZone is used to focus the first element inside it's children */
 export class AutoFocusZone extends React.Component<AutoFocusZoneProps, {}> {
-  private static _focusStack: AutoFocusZone[] = []
   private _root: { current: HTMLElement | null } = { current: null }
   private _previouslyFocusedElementOutsideTrapZone: HTMLElement
```
Comment (sophieH29) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: `_previouslyFocusedElementOutsideTrapZone ` it is not set anywhere. In FTZ, it is set usually to document.active element, so on `componentWillUnmount` we can focus it, basically does same as in Popup's action - `closeAndFocusTrigger`. I don't think we need it here.

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 3 above)
Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: Removed from here.

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 5 above)
Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: Removed from here.

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 8 above)
Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: Can you check the latest changes? `FocusTrapZone`, now returns in the render method `AutoFocusZone`. Is that what you mean with this comment?

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 13 above)
Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: This is not `UIComponent`, that's why we have this as part of `defaultProps`.

Comment (layershifter) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: https://github.com/stardust-ui/react/blob/master/packages/react/src/lib/getElementType.tsx#L41
It handled in `getElementType()`

Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: Yep, I forgot we are directly using it. Thanks!

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 14 above)
Comment (mnajdova) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: I removed it, but I am a bit concern, as this may introduce some regressions. Previously we had the following code:
```
  public componentDidMount(): void {
    const { disableFirstFocus = false } = this.props

    if (
      !this._root.current.contains(this._previouslyFocusedElementOutsideTrapZone) &&
      !disableFirstFocus
    ) {
      this._findElementAndFocusAsync()
    }
  }
```
Now we don't have the fist condition here...

Comment (sophieH29) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: yep, if look in FTZ `_previouslyFocusedElementOutsideTrapZone` is used in many places and is calculated there. Here it doesn't make sense. But we need somehow to respect this condition when `FocusTrapZone` will use `AutoFocusZone`

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx (hunk 7 above)
Comment (mnajdova) [reply docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: I can see that we have added `focusTrap` as a prop to the behavior, but I cannot see how, depending on this prop we are providing the `FocusTrapZone`. I can see that in `renderComponent` we are handling the `focusZone` prop, but I can't find anything regarding the `focusTrap`.

Comment (mnajdova) [reply docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: I added this i the Popup.. I don't like that we are handling it there, but no other idea at the moment..

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx (hunk 6 above)
Comment (mnajdova) [reply docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: You are right, I removed this selector

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+class ReactionPopup extends React.Component<any, any> {
+  state = {
+    open: false,
+  }
+
+  reactionNode: HTMLButtonElement
+
+  popup: HTMLElement
```
Comment (layershifter) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: Cruft?

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  }
+  return `${numberOfPersons} people reacted this message with a ${emojiType} emoji. Open menu to see people who reacted.`
+}
+
+class ReactionPopup extends React.Component<any, any> {
+  state = {
+    open: false,
+  }
+
+  reactionNode: HTMLButtonElement
```
Comment (layershifter) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: You can use `createRef()` there

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+      // Imported from the MenuButton prototype
+      focusNearest(this.reactionNode, e.shiftKey ? 'previous' : 'next')
+    }
+  }
+
+  handleOpenChange = (e, { open }) => {
+    this.setState({
+      open,
+    })
+  }
```
Comment (layershifter) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: Can be one-liner 😸

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+                  vertical
+                  variables={{ borderColor: 'transparent' }}
+                  onKeyDown={this.handleKeyDownOnMenu}
+                />
+              </Ref>
+            ),
+          }}
+          open={this.state.open}
+          onOpenChange={this.handleOpenChange}
+          on={'hover'}
```
Comment (layershifter) [inline docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]:
```suggestion
          on='hover'
```

File: docs/src/prototypes/MessageReactionsWithPopup/index.tsx [hunk 19]
```diff
@@ -0,0 +1,80 @@
+import * as React from 'react'
+import * as _ from 'lodash'
+import { Avatar, Chat } from '@stardust-ui/react'
+import ReactionPopup from './ReactionPopup'
+
+const reactions = [{ icon: 'thumbs up', content: '1K' }, { icon: 'thumbs down', content: 2 }]
+const reactionsWithPopup = _.map(reactions, reaction => render =>
+  render(reaction, (Component, props) => <ReactionPopup Component={Component} props={props} />),
+)
```
Comment (layershifter) [inline docs/src/prototypes/MessageReactionsWithPopup/index.tsx]: Do we need `reactionsWithPopup()`? Can we just use `{ as: ReactionPopup, icon: 'thumbs up', content: '1K' }`?

File: packages/react/src/components/Popup/Popup.tsx [hunk 20]
```diff
@@ -451,11 +455,17 @@ export default class Popup extends AutoControlledComponent<ReactProps<PopupProps
       ...popupWrapperAttributes,
     } as FocusTrapZoneProps
 
+    const autoFocusProps = {
+      ...(typeof accessibility.autoFocus === 'boolean' ? {} : accessibility.autoFocus),
+      ...popupWrapperAttributes,
+    } as FocusTrapZoneProps
```
Comment (layershifter) [inline packages/react/src/components/Popup/Popup.tsx]: Can be this part of behavior? We have there access to props

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+import { AutoFocusZoneProps } from './AutoFocusZone.types'
+import getUnhandledProps from '../../getUnhandledProps'
+import getElementType from '../../getElementType'
+import * as customPropTypes from '../../customPropTypes'
+import { callable } from '../../index'
+import Ref from '../../../components/Ref/Ref'
+
+/** AutoFocusZone is used to focus the first element inside it's children */
+export class AutoFocusZone extends React.Component<AutoFocusZoneProps, {}> {
+  private _root = React.createRef<HTMLElement>()
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```suggestion
  private root = React.createRef<HTMLElement>()
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+  private _previouslyFocusedElementInTrapZone?: HTMLElement
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```suggestion
  private previouslyFocusedElementInTrapZone?: HTMLElement
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    firstFocusableSelector: PropTypes.string,
+    disableFirstFocus: PropTypes.bool,
+    focusPreviouslyFocusedInnerElement: PropTypes.bool,
+  }
+
+  public componentDidMount(): void {
+    const { disableFirstFocus = false } = this.props
+
+    if (!disableFirstFocus) {
+      this._findElementAndFocusAsync()
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```suggestion
      this.findElementAndFocusAsync()
```

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+      { handledProps: [..._.keys(AutoFocusZone.propTypes)] },
+      this.props,
+    )
+    // TODO: Remove `as` there after the issue will be resolved:
+    // https://github.com/Microsoft/TypeScript/issues/28768
+    const ElementType = getElementType({}, this.props) as React.ComponentClass<AutoFocusZoneProps>
+
+    return (
+      <Ref innerRef={this._root}>
+        <ElementType {...unhandledProps} onFocusCapture={this._onFocusCapture}>
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]:
```suggestion
        <ElementType {...unhandledProps} onFocusCapture={this.onFocusCapture}>
```

File: packages/react/src/lib/accessibility/FocusZone/FocusTrapZone.tsx [hunk 25]
```diff
@@ -57,46 +53,38 @@ export class FocusTrapZone extends React.Component<FocusTrapZoneProps, {}> {
...
-    }
-
     this._hideContentFromAccessibilityTree()
     this._subscribeToEvents()
   }
 
   public render(): JSX.Element {
-    const { className, ariaLabelledBy } = this.props
+    const {
+      as,
```
Comment (layershifter) [inline packages/react/src/lib/accessibility/FocusZone/FocusTrapZone.tsx]: It seems that `as` is unhandled prop in this case and can be removed from `propTypes`

File: packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx (hunk 8 above)
Comment (sophieH29) [reply packages/react/src/lib/accessibility/FocusZone/AutoFocusZone.tsx]: yes 👍

File: docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx (hunk 15 above)
Comment (mnajdova) [reply docs/src/prototypes/MessageReactionsWithPopup/ReactionPopup.tsx]: Yep, removed

File: packages/react/src/components/Popup/Popup.tsx [hunk 26]
```diff
@@ -62,6 +62,9 @@ export interface PopupProps
   /** Alignment for the popup. */
   align?: Alignment
 
+  /** If this prop is used, the Popup will propagate the key down event, otherwise it won't. */
+  allowKeyDownPropagation?: boolean
```
Comment (mnajdova) [inline packages/react/src/components/Popup/Popup.tsx]: See my comment: https://github.com/stardust-ui/react/pull/987/files#r263011991 No better proposal for now..
